sdk/core/src/modify_conditions.rs

```diff
@@ -57,6 +54,14 @@ impl<'a> IfMatchCondition<'a> {
     }
 }
 
+impl<'a> crate::AddAsHeader for IfMatchCondition<'a> {
```

Comment:
At some point we'll want to take https://azure.github.io/azure-sdk/general_design.html#conditional-requests into consideration. Rarely (perhaps never in track 2) are these headers set directly, but exposed more abstractly based on a common etag struct (using the term loosely).

Reply:
I think we are on right track already. We treat the etag as opaque value. 

This is taken from an sdk example ([https://github.com/Azure/azure-sdk-for-rust/blob/536e32f68f5177c94109131f478dfbc79b61e476/sdk/cosmos/examples/document_00.rs#L186-L207](https://github.com/Azure/azure-sdk-for-rust/blob/536e32f68f5177c94109131f478dfbc79b61e476/sdk/cosmos/examples/document_00.rs#L186-L207):

```rust
    // The document can be no longer there so the result is
    // an Option<Document<T>>
    if let GetDocumentResponse::Found(document) = get_document_response {
        // Now, for the sake of experimentation, we will update (replace) the
        // document created. We do this only if the original document has not been
        // modified in the meantime. This is called optimistic concurrency.
        // In order to do so, we pass to this replace_document call
        // the etag received in the previous get_document. The etag is an opaque value that
        // changes every time the document is updated. If the passed etag is different in
        // CosmosDB it means something else updated the document before us!
        let replace_document_response = collection_client
            .replace_document()
            .with_document_id(&doc.document.id)
            .with_partition_keys(&(&doc.document.id).into())
            .with_if_match_condition(IfMatchCondition::Match(&document.etag))
            .execute_with_document(&doc)
            .await?;
        println!(
            "replace_document_response == {:#?}",
            replace_document_response
        );
    }
```

From the link you mentioned I think we miss one of the requisites:



status | Item | info
-- | -- | --
🟢 | YOU SHOULD accept a conditions parameter (which takes an enumerated type) on service methods that allow a conditional check on the service. | We have this (at least in CosmosDB). For other crates we should check and make sure they are added.
🟢 | YOU SHOULD accept an additional boolean or enum parameter on service methods as necessary to enable conditional checks using ETag | We have this (as seen in the aforementioned example).
🟢 | YOU SHOULD include the ETag field as part of the object model when conditional operations are supported | We have this (at least in CosmosDB). For other crates we should check and make sure they are added.
🔴 | YOU SHOULD NOT throw an error when a 304 Not Modified response is received from the service, unless such errors are idiomatic to the language. | This does not happen right now, we should address it (we erroneously return an error). We should probably make sure to accept this return code as valid only if a precondition has been set in the request.
🟡 | YOU SHOULD throw a distinct error when a 412 Precondition Failed response or a 409 Conflict response is received from the service due to a conditional check. | Right now errors include the specific HTTP code so we are compliant. We should make errors more explict though.

I would also like to make etag more opaque (using a struct or a trait), like the work @rylev is doing for the other options.

Reply:
In libraries where we've used etags, we really don't expose it that way either. I can't say I'm a fan of it, but we typically take an object for a create_or_modify-type method and a bool that says only_if_unchanged. E.g. in Cognitive Search that I wrote according to spec: https://github.com/Azure/azure-sdk-for-net/blob/2a29edb99ac173414de34d1b78865deb04607791/sdk/search/Azure.Search.Documents/src/Indexes/SearchIndexClient.cs#L414-L428

When then do this consistently across languages, though names, how it's passed (e.g. kwargs for Python) change.

Again: it's not something we need to do now, but I do worry the further we diverge from something idiomatic-yet-consistent, the harder it will be remember to change later. Some patterns will and should be different (like builders in Rust or Java vs. instantiation in C#, JS, and Python) but method signatures are generally pretty consistent (like defaulted parameter values for C#, overloads for Java, and kwargs/options bags for Python and JS).

Reply:
Hi @heats, 
thank you for helping us to go in the right direction!

This is taken from the official C# sdk:
(from [https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.requestoptions.accesscondition?view=azure-dotnet#Microsoft_Azure_Documents_Client_RequestOptions_AccessCondition](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.requestoptions.accesscondition?view=azure-dotnet#Microsoft_Azure_Documents_Client_RequestOptions_AccessCondition)):

```c#
// If ETag is current, then this will succeed. Otherwise the request will fail with HTTP 412 Precondition Failure
await client.ReplaceDocumentAsync(
    readCopyOfBook.SelfLink,
    new Book { Title = "Moby Dick", Price = 14.99 },
    new RequestOptions
    {
        AccessCondition = new AccessCondition
        {
            Condition = readCopyOfBook.ETag,
            Type = AccessConditionType.IfMatch
        }
     });
```

As reference, this is the same code in this SDK:

```rust
let replace_document_response = collection_client 
         .replace_document() 
         .with_document_id(&readCopyOfBook.id) 
         .with_partition_keys(&(&readCopyOfBook.document.id).into()) 
         .with_if_match_condition(IfMatchCondition::Match(&readCopyOfBook.etag)) 
         .execute_with_document(&new_doc) 
         .await?; 
```

The only difference I see is that Rust treats id and partition keys as separate entities while C# uses the self link. We can change that (as it makes sense for a replace) or allow both. 
The `etag` code is basically the same.

What do you think?

Reply:
Shall we move this to an issue? I'd hate for this discussion to get lost.